### ISSUE TITLE
Fan-out pipelines are slow

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#7813]
## Context

Thanks to the magic of GraphQL, it's very easy to build a "fan-out pipelines": pipelines where one step returns an array of objects, and the next step applies the same function to all objects in the array.

For example, given this module code:

```golang
package main

type Foo struct {}

// Receive a list of messages, and produce a corresponding list of files
func (foo Foo) Files(messages []string) []*File {
  files := make([]*File, len(messages))
  for i := range messages {
    files[i] = dag.
      Directory().
      WithNewFile("message.txt", messages[i]).
      File("message.txt")
  }
  return files
}
```

We can call a simple fan-out pipeline:

```console
dagger call \
  files --messages hello,bonjour,hola,ciao \
  contents
```

The output will be:

```console
hello
bonjour
hola
ciao
```

## Problem

Fan-out pipelines are great, but for any non-trivial implementation, they are very slow. This seems to be caused by very inefficient implementation, specifically: each object in the array seems to be queried serially, instead of concurrently.

For example, here is a variation of the previous example, modified to simulate an "expensive" operation:

```go
package main

import (
	"context"
	"time"
)

type Foo struct{}

// Receive a list of messages, and produce a corresponding list of files
func (foo Foo) Files(messages []string) []*SlowFile {
	files := make([]*SlowFile, len(messages))
	for i := range messages {
		files[i] = &SlowFile{
			File: dag.
				Directory().
				WithNewFile("message.txt", messages[i]).
				File("message.txt"),
		}
	}
	return files
}

type SlowFile struct {
	File *File // +private
}

func (f *SlowFile) Contents(ctx context.Context) (string, error) {
	time.Sleep(1 * time.Second)
	return f.File.Contents(ctx)
}

func (f *SlowFile) Size(ctx context.Context) (int, error) {
	time.Sleep(1 * time.Second)
	return f.File.Size(ctx)
}
```

```console
$ dagger -v call \
  files --messages hello,bonjour,hola,ciao,salut \
  contents

✔ connect 0.5s
  ✔ starting engine 0.4s
    ✔ create 0.4s
✔ initialize 0.3s
✔ foo: Foo! 0.0s
✔ Foo.files(messages: ["hello", "bonjour", "hola", "ciao", "salut"]): [FooSlowFile!]! 0.2s
✔ prepare 0.0s
✔ Void.contents: String! 1.3s
✔ Void.contents: String! 1.2s
✔ Void.contents: String! 1.2s
✔ Void.contents: String! 1.3s
✔ Void.contents: String! 1.2s

hello
bonjour
hola
ciao
salut
dagger -v call files --messages hello,bonjour,hola,ciao,salut contents  0.33s user 0.23s system 7% cpu 7.533 total
```

Note the total duration: 7.5 seconds! This illustrated the huge performance impact of querying each object in the array serially, regardless of actual CPU cost.

## Solution

Make fan-out pipelines less punitively slow for non-trivial implementations. Possibly by making sure all items in an array are queried concurrently, rather than serially.

**Comment by vito** [comment]
While we're in the area, we should fix this side issue @helderco pointed out from the output above:

```
✔ Void.contents: String! 1.3s
✔ Void.contents: String! 1.2s
✔ Void.contents: String! 1.2s
✔ Void.contents: String! 1.3s
✔ Void.contents: String! 1.2s
```

Here the TUI is probably filling in `Void` as a wonky fallback because it's confused by functions called against list values (for whatever reason).

As for the _behavior_ issue this might be a simple change in DagQL to resolve enumerable return values in parallel.

**Comment by helderco** [comment]
This looks like a duplicate of https://github.com/dagger/dagger/issues/7813

**Comment by shykes** [comment]
> This looks like a duplicate of https://github.com/dagger/dagger/issues/7813

Typo?

**Comment by helderco** [comment]
> Typo?

Yeah, sorry 😅 I meant this:
- https://github.com/dagger/dagger/issues/7353

**Comment by samalba** [comment]
Shouldn't the concurrency be managed at the language-level in that case (go routines with a waitgroup)?

**Comment by valorl** [comment]
I've just ran into this myself as well, forcing me to create aggregate `build-all`, `publish-all` functions instead of using chaining (e.g. `find-applications (array) -> build -> publish`) which was a lot more elegant on the CLI.

> Shouldn't the concurrency be managed at the language-level in that case (go routines with a waitgroup)?

From what I can tell this is currently the only way currently to run the fan-out concurrently and also what you need to do if you're calling the functions from code/SDK. My understanding is that if you e.g. `Sync()` in separate go-routines, you're using the language to send separate sync requests to the engine and wait on them concurrently. However, the CLI (and presumably the engine/GraphQL) allows you to chain an array result (`[]T`) into a scalar function (`f(T)`), automatically calling it for each result, which is really convenient, but unfortunately happens sequentially.

**Comment by M-Pixel** [comment]
> Shouldn't the concurrency be managed at the language-level in that case (go routines with a waitgroup)?

This is exactly what I've done in my dotnet SDK since the 0.9 days.  Every parameter in a query that is an object instead of a scalar gets dispatched in parallel using C#'s TPL (equivalent to Goroutines).

**Comment by helderco** [comment]
Closing as a duplicate of https://github.com/dagger/dagger/issues/7353. Feel free to reopen if not the case.
